Consensus protocol design doc (#100)
tomarrell/lbadd: doc/internal/consensus/Consensus-Protocol.md

# The consensus

Before talking about consensus, we need to discuss some logistics based on how the systems can co-exist.

> * Communication: Distributed systems need a method to communicate between each other. Remote Procedure Calls is the mechanism using which a standalone server can talk to another. The existing `network` layer of the database will handle all the communication between servers. 

Reviewer: However, only the communication. Messages etc. need to be specified and are part of the Raft implementation.

> * Security: Access control mechanisms need to be in place to decide on access to functions in the servers based on their state (leader, follower, candidate)

Reviewer: Not a raft problem. We need to implement an access control layer that is configured with SQL statements, such as in common databases. They are communicated over raft (replicating statements), but not part of the consensus protocol however.

Author: What I meant were raft functions.

There are variables that only the leader can maintain.

> * Routing to leader: One of the issues with a varying leader is for the clients to know which IP address to contact for the service. We can solve this problem by advertising any/all IPs of the cluster and simply forward this request to the current leader; OR have a proxy that can forward the request to the current leader wheneve the requests come in. (Section client interaction of post has another approach which works too)

Reviewer: As this is a design doc, please make a decision, and state, why you made that decision, and why you didn't choose the other solution.

Personally, I probably would make the discovery of all other nodes part of the cluster login handshake, but it's your decision.

Author: The cluster login handshake is an initial setup problem, yes.

But here since the leaders keep changing, we need another method to figure out who the leader is.

I'll stick to the Client providing the IP one, seems better.

* The servers will be implemented in the `interal/node` folders which will import the raft API and perform their functions.

Maintaining consensus is one of the major parts of a distributed system. To know to have achieved a stable system, we need the following two parts of implementation.

## The Raft protocol

The raft protocol will be implemented in `internal/raft` and will implement APIs that each node can call.

> Raft is an algorithm to handle replicated log, and we maintain the "log" of the SQL stmts applied on a DB and have a completely replicated cluster.

Reviewer: ```suggestion
Raft is an algorithm to handle replicated log, and we maintain the "log" of the SQL statements applied on a DB and have a completely replicated cluster.
```

#### General Implementation rules:
* All RPC calls are done parallely to obtain the best performance.
* Request retries are done in case of network failures.
* Raft does not assume network preserves ordering of the packets.

> A raft server may be in any of the 3 states; leader, follower or candidate. All requests are serviced through the leader and it then decides how and if the logs must be replicated in the follower machines. The raft protocol has 3 almost independent modules:

Reviewer: Independent modules means that they live in `internal/raft/{election,replication,safety}`? Or do they all live in `internal/raft`?

Author: They will be in a single package `raft` just because I think it'll be easier to use.
By separate modules, I meant they don't interact with each other; though they are in the same package.

Author: What do you think?

Reviewer: I think that's fine. However, if that bloats the package too much (which I don't think), we should think about moving if possible.

1. Leader Election
2. Log Replication
3. Safety

A detailed description of all the modules follow:

### Leader Election

#### Spec

> * Startup: All servers start in the follower state and begin by requesting votes to be elected as a leader.

Reviewer: What happens with the first node in the cluster?
Does he detect that there are no others and therefore falls into the leader state?
How do we handle this case?

Author: Im not exactly sure. Since there are no other nodes, the node can vote for itself and become the leader.

But we need to discuss this.

Maybe we should start all nodes simultaneously?

Reviewer: voting for itself makes sense I guess...

* Pre-election: The server increments its `currentTerm` by one, changes to `candidate` state and sends out `RequestVotes` RPC parallely to all the peers. 

> * Vote condition: FCFS basis. If there was no request to the server, it votes for itself (read 3.6 and clear out when to vote for itself)

Reviewer: 3.6 where? Maybe add a link if possible

Author: That was a note for myself while reading ^^'
Removed.

* Election timeout: A preset time for which the server waits to see if a peer requested a vote. It is randomly chosen between 150-300ms.
* Election is repeated after an election timeout until:
  1. The server wins the election
  2. A peer establishes itself as leader.
  3. Election timer times out or a split vote occurs (leading to no leader) and the process will be repeated.
 * Election win: Majority votes in the term. (More details in safety) The state of the winner is now `Leader` and the others are `Followers`.

>  * The term problem: Current terms are exchanged when-ever servers communicate; if one server’s current term is smaller than the other’s, then it updatesits current term to the larger value. If a candidate or leader discovers that its term is out of date,it immediately reverts to follower state. If a server receives a request with a stale term number, itrejects the request.

Reviewer: ```suggestion
 * The term problem: Current terms are exchanged when-ever servers communicate; if one server’s current term is smaller than the other’s, then it updates its current term to the larger value. If a candidate or leader discovers that its term is out of date,it immediately reverts to follower state. If a server receives a request with a stale term number, itrejects the request.
```

>  * The term problem: Current terms are exchanged when-ever servers communicate; if one server’s current term is smaller than the other’s, then it updatesits current term to the larger value. If a candidate or leader discovers that its term is out of date,it immediately reverts to follower state. If a server receives a request with a stale term number, itrejects the request.

Reviewer: ```suggestion
 * The term problem: Current terms are exchanged when-ever servers communicate; if one server’s current term is smaller than the other’s, then it updatesits current term to the larger value. If a candidate or leader discovers that its term is out of date, it immediately reverts to follower state. If a server receives a request with a stale term number, it rejects the request.
```

What is an 'out of date term'?

Author: Out of date terms are those which are _not the current term_ basically. 
I'll maybe add that to the doc.

 * Maintaining leaders reign: The leader sends `heartbeats` to all servers to establish its reign. This also checks whether other servers are alive based on the response and informs other servers that the leader still is alive too. If the servers do not get timely heartbeat messages, they transform from the `follower` state to `candidate` state.

>  * Transition from working state to Election happens when a leader fails.

Reviewer: Working state? `Follower`?

Author: Actually it means, leaders normal working.

Author: Made clear in the doc.

>  * Maintaining sanity: While waiting for votes, if a `AppendEntriesRPC` is received by the server, and the term of the leader is greater than of equal to the "waiter"'s term, the leader is considered to be legitimate and the waiter becomes a follower of the leader. If the term of the leader is lesser, it is rejected.

Reviewer: ```suggestion
 * Maintaining sanity: While waiting for votes, if a `AppendEntriesRPC` is received by the server, and the term of the leader is greater than of equal to the waiting node's term, the leader is considered to be legitimate and the waiter becomes a follower of the leader. If the term of the leader is lesser, it is rejected.
```

>  * Maintaining sanity: While waiting for votes, if a `AppendEntriesRPC` is received by the server, and the term of the leader is greater than of equal to the "waiter"'s term, the leader is considered to be legitimate and the waiter becomes a follower of the leader. If the term of the leader is lesser, it is rejected.

Reviewer: ```suggestion
 * Maintaining sanity: While waiting for votes, if a `AppendEntriesRPC` is received by the server, and the term of the leader is greater than of equal to the "waiter"'s term, the leader is considered to be legitimate and the waiter becomes a follower of the leader. If the term of the leader is smaller, it is rejected.
```

>  * The split vote problem: Though not that common, split votes can occur. To make sure this doesnt continue indefinitely, election timeouts are randomised, making the split votes less probable.

Reviewer: ```suggestion
 * The split vote problem: Though not that common, split votes can occur. To make sure this doesn't continue indefinitely, election timeouts are randomised, making the split votes less probable.
```

* Repeating `AppendEntries`: `AppendEntriesRPC` are repeated indefinitely until all followers eventually store all log entries.
* Log entry storage: Log entries are a queue of state machine commands which are applied to that particular state machine. Log entries are associated with a term number to indicate the term of application of that log along with an integer index to identify a particular logs position.
* A committed entry: When a leader decides that the log entry is safe to apply to other state machines, that entry is called committed. All committed entries are durable and _will eventually be executed_ by all state machines.
* An entry -> Committed entry: A log entry is called committed once its replicated on the majority of the servers in the cluster. Once an entry is committed, it commits all the previous entries in the leaders log, including the entries created by the previous leaders.
* The  leader keeps track of the highest known index that it knows is committed and it is included in all the future `AppendEntriesRPC` (including heartbeats) to inform other servers.

> * Theres some issue about log committing - "A log entry is committed once the leader that createdthe entry has replicated it on a majority of the servers" and " Once a follower learns that a log entry is committed, it applies theentry to its local state machine (in log order)." are not clear whether replicating and applying to state machine are the same. If they are its kind of a contradiction, else "aplication" can mean executing the STMT in the DB in our case.

Reviewer: ```suggestion
* Theres some issue about log committing - "A log entry is committed once the leader that createdthe entry has replicated it on a majority of the servers" and " Once a follower learns that a log entry is committed, it applies theentry to its local state machine (in log order)." are not clear whether replicating and applying to state machine are the same. If they are its kind of a contradiction, else "application" can mean executing the STMT in the DB in our case.
```

> * Theres some issue about log committing - "A log entry is committed once the leader that createdthe entry has replicated it on a majority of the servers" and " Once a follower learns that a log entry is committed, it applies theentry to its local state machine (in log order)." are not clear whether replicating and applying to state machine are the same. If they are its kind of a contradiction, else "aplication" can mean executing the STMT in the DB in our case.

Reviewer: ```suggestion
* Theres some issue about log committing - "A log entry is committed once the leader that createdthe entry has replicated it on a majority of the servers" and " Once a follower learns that a log entry is committed, it applies theentry to its local state machine (in log order)." are not clear whether replicating and applying to state machine are the same. If they are its kind of a contradiction, else "aplication" can mean executing the statement in the database in our case.
```

* Log matching property: 

>   * If two entries in different logs have the same index and term, then they store the same com-mand.

Reviewer: ```suggestion
  * If two entries in different logs have the same index and term, then they store the same command.
```

>   * If two entries in different logs have the same index and term, then the logs are identical in allpreceding entries.

Reviewer: ```suggestion
  * If two entries in different logs have the same index and term, then the logs are identical in all preceding entries.
```

> * When sending an AppendEntriesRPC, the leader includes the index and term of the entry in its log that immediately precedes thenew entries. If the follower does not find an entry in its log with the same index and term, then itrefuses the new entries. This helps in log matching. Which implies, a successful `AppendEntries` RPC means a synced log.

Reviewer: ```suggestion
* When sending an AppendEntriesRPC, the leader includes the index and term of the entry in its log that immediately precedes the new entries. If the follower does not find an entry in its log with the same index and term, then it refuses the new entries. This helps in log matching. Which implies, a successful `AppendEntriesRPC` means a synced log.
```

> * Leader crashes inducing inconsistencies in logs: In Raft, the leader handles inconsistencies by forcing the followers’ logs to duplicate its own (the leader's). To do so, the leader must find the latest log entrywhere  the  two  logs  agree,  delete  any  entries  in  the  follower’s  log  after  that  point,  and  send  thefollower all of the leader’s entries after that point. All of these actions happen in response to theconsistency check performed by AppendEntries RPCs. Meaning, the leader checks for the consistency by maintaining a `nextIndex` value and dropping it down and sending `AppendEntries` RPC (which does the consistency check and fails unless they're same) until a success is returned. These "check `AppendEntries` can be empty to save BandWidth(BW). The follower can also help here by sending the smallest agreeing index in the first RPC instead of the leader probing until it reaches the index. 

Reviewer: ```suggestion
* Leader crashes inducing inconsistencies in logs: In Raft, the leader handles inconsistencies by forcing the followers’ logs to duplicate its own (the leader's). To do so, the leader must find the latest log entry where  the  two  logs  agree,  delete  any  entries  in  the  follower’s  log  after  that  point,  and  send  the follower all of the leader’s entries after that point. All of these actions happen in response to the consistency check performed by AppendEntries RPCs. Meaning, the leader checks for the consistency by maintaining a `nextIndex` value and dropping it down and sending `AppendEntriesRPC` (which does the consistency check and fails unless they're same) until a success is returned. These "check `AppendEntries` can be empty to save BandWidth(BW). The follower can also help here by sending the smallest agreeing index in the first RPC instead of the leader probing until it reaches the index. 
```

# The consensus

Before talking about consensus, we need to discuss some logistics based on how the systems can co-exist.

> * Communication: Distributed systems need a method to communicate between each other. Remote Procedure Calls is the mechanism using which a standalone server can talk to another. The standard Go package [RPC](https://golang.org/pkg/net/rpc/) serves us the purpose. 

Reviewer: How does RPC perform (code-wise) in an 1-to-many scenario?

Author: This is changed.

* Security: Access control mechanisms need to be in place to decide on access to functions in the servers based on their state (leader, follower, candidate)
* Routing to leader: One of the issues with a varying leader is for the clients to know which IP address to contact for the service. We can solve this problem by advertising any/all IPs of the cluster and simply forward this request to the current leader; OR have a proxy that can forward the request to the current leader wheneve the requests come in. (Section client interaction of post has another approach which works too)

> * The servers will be implemented in the `interal/master` or `internal/worker` folders which will import the raft API and perform their functions.

Reviewer: `internal/node` 😄

Author: Changed, older version.

* Successful log: When all logs have been applied successfully to all follower machines, the leader applies the entry to its state machine and returns the result to the client.
* Repeating `AppendEntries`: `AppendEntriesRPC` are repeated indefinitely until all followers eventually store all log entries.
* Log entry storage: Log entries are a queue of state machine commands which are applied to that particular state machine. Log entries are associated with a term number to indicate the term of application of that log along with an integer index to identify a particular logs position.
* Committed entry: A log entry is called committed once its replicated on the majority of the servers in the cluster. Once an entry is committed, it commits all the previous entries in the leaders log, including the entries created by the previous leaders.
* The  leader keeps track of the highest known index that it knows is committed and it is included in all the future `AppendEntriesRPC` (including heartbeats) to inform other servers.

> * Theres some issue about log committing - paragraph says its committed when its applied everywhere and also says its applied everywhere once its committed. 

Reviewer 2: what issue in log commiting, I think in doc we should put how we are going to design and what we are going to do for given problem rather talk about problem. We can discuss find it's solution or working solution

Author: Outdated.

* Log matching property: 
  * If two entries in different logs have the same index and term, then they store the same com-mand.
  * If two entries in different logs have the same index and term, then the logs are identical in allpreceding entries.
 * 
* Add more from section 3.6

#### Implementation

### Safety

## A strict testing mechanism

The testing mechanism to be implemented will enable us in figuring out the problems existing in the implementation leading to a more resilient system.
We have to test for the following basic failures:
1. Network partitioning.
2. Un-responsive peers.
3. Overloading peer.
4. Corruption of data in transit.

## Graceful handling of failures

Accepting failures exist and handling them gracefully enables creation of more resilient and stable systems. Having _circuit breakers_, _backoff mechanisms in clients_ and _validation and coordination mechanisms_ are some of the pointers to be followed. 

## Running Lbadd on Raft

* Background: Raft is just a consensus protocol that helps keep different database servers in sync. We need methods to issue a command and enable the sync between servers.

> * Logistics: The `AppendEntriesRPC` will have the command to be executed by the client. This command goes through the leader, is applied by all the followers and then committed by the leader. Thus ensuring an in-sync distributed database.

Reviewer 2: This command is invoked by leader. Each client request contains a command to be executed by the replicated state machines. The leader append that command to its log entries and then issues AppendEntriesRPC to rest of the follower. It is also use as heartbeat by the leader

So command doesn't go through the leader, is invoked and also not executed by the client

Author: Everything goes through the leader and no-one else.
The log in `AppendEntries` RPC will have a compiled version of the SQL stmt that the follower nodes will apply.

Before talking about consensus, we need to discuss some logistics based on how the systems can co-exist.

* Communication: Distributed systems need a method to communicate between each other. Remote Procedure Calls is the mechanism using which a standalone server can talk to another. The existing `network` layer of the database will handle all the communication between servers. However, the messages to be passed are decided by the raft layer.
* Security: Access control mechanisms need to be in place to decide on access to functions in the servers based on their state (leader, follower, candidate)

> * Routing to leader: One of the issues with a varying leader is for the clients to know which IP address to contact for the service. We can solve this problem by advertising any/all IPs of the cluster and the client returns the IP of the leader if its not the leader.

Reviewer: I don't understand. Any node knows all IPs and can answer that question. Why should follower nodes only respond with the leader IP?

Author: At any point of time when raft is in the "working phase" (which is the leader is up, no election is happenning), there are only 2 kinds of nodes; the leader and the follower. So if the client doesn't hit the leader, it means it hit the follower.

 * Maintaining leaders reign: The leader sends `heartbeats` to all servers to establish its reign. This also checks whether other servers are alive based on the response and informs other servers that the leader still is alive too. If the servers do not get timely heartbeat messages, they transform from the `follower` state to `candidate` state.
 * Transition from leader's normal working state to Election happens when a leader fails.
 * Maintaining sanity: While waiting for votes, if a `AppendEntriesRPC` is received by the server, and the term of the leader is greater than of equal to the waiting node's term, the leader is considered to be legitimate and the waiter becomes a follower of the leader. If the term of the leader is lesser, it is rejected.
 * The split vote problem: Though not that common, split votes can occur. To make sure this doesn't continue indefinitely, election timeouts are randomised, making the split votes less probable.
 

#### Implementation

> * The implementation of leader election will span over `raft/leaderElection.go`, request votes over `raft/requestVotes.go` and append entries over `appendEntries.go`.

Reviewer: ```suggestion
* The implementation of leader election will span over `raft/leader_election.go`, request votes over `raft/request_votes.go` and append entries over `append_entries.go`.
```
Go uses snake case

Author: Oh? Alright!

* Request retries are done in case of network failures.
* Raft does not assume network preserves ordering of the packets.
* The `raft/raft.go` file has all the state parameters and general functions that raft uses.
* A raft server is implemented as:
```
type simpleServer struct {
	node          *Node
	cluster       Cluster
	onReplication ReplicationHandler
	log           zerolog.Logger
}

type Node struct {
	State string

	PersistentState     *PersistentState
	VolatileState       *VolatileState
	VolatileStateLeader *VolatileStateLeader
}

> 

Reviewer: why is there implementation in a design doc?

Author: Well if I add the API, they struct's would be a good reference to have, was my thought.

Reviewer: Keep in mind that you are writing a design doc, not an implementation guide. Such a concrete struct may have heavy implications on the rest of the implementation. The point of a design doc is, to just outline the components and how they interact, and leave the concrete implementation to the developer.

* Leader crashes inducing inconsistencies in logs: In Raft, the leader handles inconsistencies by forcing the followers’ logs to duplicate its own (the leader's). To do so, the leader must find the latest log entry where  the  two  logs  agree,  delete  any  entries  in  the  follower’s  log  after  that  point,  and  send  the follower all of the leader’s entries after that point. All of these actions happen in response to the consistency check performed by AppendEntries RPCs. Meaning, the leader checks for the consistency by maintaining a `nextIndex` value and dropping it down and sending `AppendEntriesRPC` (which does the consistency check and fails unless they're same) until a success is returned. These "check `AppendEntries` can be empty to save BandWidth(BW). The follower can also help here by sending the smallest agreeing index in the first RPC instead of the leader probing until it reaches the index. 
* Leader append-only property: A leader never overwrites or deletes entries in its own log.

#### Implementation

> * Log replication implementation will span over the `raft/logReplication.go` file.

Reviewer: snake case
```suggestion
* Log replication implementation will span over the `raft/log_replication.go` file.
```

* A committed entry: When a leader decides that the log entry is safe to apply to other state machines, that entry is called committed. All committed entries are durable and _will eventually be executed_ by all state machines.
* An entry -> Committed entry: A log entry is called committed once its replicated on the majority of the servers in the cluster. Once an entry is committed, it commits all the previous entries in the leaders log, including the entries created by the previous leaders.
* The  leader keeps track of the highest known index that it knows is committed and it is included in all the future `AppendEntriesRPC` (including heartbeats) to inform other servers.

> * Theres some issue about log committing - "A log entry is committed once the leader that created the entry has replicated it on a majority of the servers" and " Once a follower learns that a log entry is committed, it applies theentry to its local state machine (in log order)." are not clear whether replicating and applying to state machine are the same. If they are its kind of a contradiction, else "application" can mean executing the STMT in the DB in our case.

Reviewer: resolve the issue and add one or more solutions to this doc
